Sign in

YassenAhmed-gif opened the pull request.

YassenAhmed-gif committed:
feat: Implement user sign-in functionality

This commit introduces the necessary components for user authentication, specifically focusing on the sign-in process.

Key changes include:
- Added `AuthRemoteDataSourceImpl` to handle API requests for authentication.
- Implemented `ApiServices` for defining authentication-related API endpoints.
- Updated `pubspec.lock` with new dependencies.
- Created `MainLayout` as a basic structure for the main application screen.
- Defined `AuthRemoteDataSource` interface for abstracting authentication data sources.
- Implemented `AuthRepoImpl` to manage authentication-related data operations.
- Defined `AuthRepo` interface for abstracting authentication repositories.
- Updated `DioModule` for API client configuration.
- Modified `main.dart` to set the initial route to the sign-in screen.
- Added new size constants in `sizes.dart`.
- Added new route constants in `app_routes.dart`.
- Added API constants in `api_constants.dart`.
- Updated app constants in `app_constants.dart`.
- Created `SigninRequestDto` for structuring sign-in request data.
- Updated `route_generator.dart` to include new routes for authentication and main layout.
- Created `SigninResponseDto` for structuring sign-in response data.
- Enhanced `ApiResult` to include a `when` method for handling success and error cases.
- Created `SigninResponseEntity` to represent the sign-in entity.
- Added `CustomElevatedButton` widget.
- Corrected a color constant in `app_colors.dart`.
- Updated `StorageInterface` for data storage abstraction.
- Added English and Arabic translations for authentication-related UI elements.
- Created `SigninUseCase` for handling the sign-in business logic.
- Updated `SecureStorageImpl` for secure data storage.
- Created `ForgetPassword` screen.
- Created `SigninViewModel` and `SignInState` for managing the sign-in screen's state.
- Implemented the `SigninScreen` UI.
- Created the `SignUpScreen` UI.

YassenAhmed-gif committed:
Refactor: Update sign-in logic and state management

This commit refactors the sign-in feature by:
- Simplifying the `SigninUseCase` to directly return the result from the `authRepo`.
- Removing the `when` method from `ApiResult` and handling success/error cases directly in the `SigninViewModel`.
- Updating `SigninViewModel` to use a single `SignInState` class with `copyWith` for state updates, replacing separate state classes.
- Modifying `SigninScreen` to adapt to the new state management in `SigninViewModel`.
- Adjusting `AuthRepoImpl` to directly return the result from the `_authRemoteDataSource`.

YassenAhmed-gif committed:
feat: Implement token storage after sign-in

This commit introduces the functionality to store the authentication token locally after a successful user sign-in.

Key changes include:
- Added `writeToken` method to `AuthLocalDataSource` interface and its implementation `AuthLocalDataSourceImpl` to handle saving the token using secure storage.
- Updated `AuthRepoImpl` to call `writeToken` from `AuthLocalDataSource` upon successful sign-in to store the received token.
- Added a `token` constant to `AppConstants`.
- Introduced a `_token` variable in `SigninScreen` (currently unused but likely for future token handling).

wasimGhoniem commented on lib/Feature/auth/api/model/requests/sign_in_request_dto.dart:
```diff
@@ -0,0 +1,26 @@
+@JsonSerializable()
+class SigninRequestDto {
+  @JsonKey(name: "email")
+  final String? email;
```
this is a request class, could you please make the variables required instead of nullable.

wasimGhoniem commented on lib/Feature/auth/api/model/responses/sign_in_response_dto.dart:
```diff
@@ -0,0 +1,102 @@
+}
+
+@JsonSerializable()
+class User {
```
Placing the User class in its own separate file (for example, user_dto.dart) will allow you to import and use it across different parts of your project

wasimGhoniem commented on lib/Feature/auth/data/dataSources/auth_remote_data_source.dart:
```diff
@@ -1 +1,9 @@
+
+abstract interface class AuthRemoteDataSource {
+  Future<ApiResult<SigninResponseEntity>> signin({
+    required String email,
```
Instead of passing email and password as separate String parameters to the signin method, please use a dedicated request class (for example, SigninRequestEntity).

YassenAhmed-gif committed:
feat: Implement "Remember Me" functionality

This commit introduces the "Remember Me" feature to the sign-in process.

Key changes include:
- Added `rememberMeChecked` parameter to `SigninUseCase`, `AuthRepo`, and `AuthRepoImpl` to handle the "Remember Me" option.
- Updated `AuthLocalDataSource` and `AuthLocalDataSourceImpl` to include methods for setting and getting the "Remember Me" status.
- Added `rememberMe` constant in `app_constants.dart`.
- Modified `SigninViewModel` to pass the `rememberMeChecked` value.
- Updated `SigninScreen` to include a "Remember Me" checkbox and pass its state during sign-in.
- In `AuthRepoImpl`, if "Remember Me" is checked, the preference is saved in local storage.
- Updated `main.dart` to determine the initial route based on the "Remember Me" status. If it's true, the user is navigated to the main layout; otherwise, to the sign-in screen.
- Removed unused `homeRoute` from `app_routes.dart`.

wasimGhoniem commented on lib/Feature/auth/data/repositories/auth_repo_impl.dart:
```diff
@@ -1,5 +1,27 @@
+
+  @override
+  Future<ApiResult<SigninResponseEntity>> signin({
+    required String email,
```
Instead of passing email and password as separate String parameters to the signin method, please use a dedicated request class (for example, SigninRequestEntity).

wasimGhoniem commented on lib/Feature/auth/domain/entity/sign_in_entity.dart:
```diff
@@ -0,0 +1,39 @@
+    this.token,
+  });
+}
+
```
Placing the User class in its own separate file (for example, user_entity.dart) will allow you to import and use it across different parts of your project

YassenAhmed-gif replied:
I moved UserEntity to seperated file called 'user_entity.dart'

wasimGhoniem commented on lib/Feature/auth/domain/repositories/auth_repo.dart:
```diff
@@ -1 +1,9 @@
+
+abstract interface class AuthRepo {
+  Future<ApiResult<SigninResponseEntity>> signin({
+    required String email,
```
use request entity

wasimGhoniem commented on lib/Feature/auth/domain/use_cases/sign_in_use_case.dart:
```diff
@@ -0,0 +1,28 @@
+  });
+
+  Future<ApiResult<SigninResponseEntity>> invoke({
+    required String email,
```
use request entity

wasimGhoniem commented on lib/Feature/auth/domain/use_cases/sign_in_use_case.dart:
```diff
@@ -0,0 +1,28 @@
+      password: password
+    );
+
+    return result.when(
```
just return   await authRepo.signin

wasimGhoniem commented on lib/Feature/auth/presentation/pages/sign_in_screen.dart:
```diff
@@ -0,0 +1,216 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:flower_e_commerce_app/core/helpers/routing_extensions.dart';
+import 'package:flower_e_commerce_app/core/utils/Constants/sizes.dart';
+
```
remove space

wasimGhoniem commented on lib/Feature/auth/presentation/pages/forget_password.dart:
```diff
@@ -0,0 +1,18 @@
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(title: const Text('Forget Password')),
```
do not push this file

wasimGhoniem commented on lib/Feature/auth/presentation/pages/sign_in_screen.dart:
```diff
@@ -0,0 +1,216 @@
+
+class _SigninScreenState extends State<SigninScreen> {
+  final SigninViewModel _viewModel = getIt<SigninViewModel>();
+  final _formKey = GlobalKey<FormState>();
```
logic and data (such as form controllers, form key, and _rememberMe) to be managed by the SigninViewModel rather than directly inside the ui. This will help separate UI logic from business/data logic, making the code cleaner and more maintainable.

wasimGhoniem commented on lib/Feature/auth/presentation/pages/sign_in_screen.dart:
```diff
@@ -0,0 +1,216 @@
+              crossAxisAlignment: CrossAxisAlignment.start,
+              children: [
+                const SizedBox(height: AppSizes.spaceBetwwenItems_16),
+                // Custom Back Button
```
remove comments

wasimGhoniem commented on lib/Feature/auth/presentation/pages/sign_in_screen.dart:
```diff
@@ -0,0 +1,216 @@
+                          ),
+                          hintText: LocaleKeys.EmailHint,
+                        ),
+                        validator: (value) {
```
use validations

wasimGhoniem commented on lib/Feature/auth/presentation/pages/sign_in_screen.dart:
```diff
@@ -0,0 +1,216 @@
+                              Checkbox(
+                                value: _rememberMe,
+                                onChanged: (val) {
+                                  setState(() {
```
do not use set state handle it with cubit

wasimGhoniem commented on lib/Feature/auth/presentation/pages/sign_up_screen.dart:
```diff
@@ -0,0 +1,19 @@
+import 'package:flower_e_commerce_app/core/Widgets/custom_app_bar.dart';
```
do not push this

wasimGhoniem commented on lib/Feature/auth/presentation/viewModel/signin/sign_in_state.dart:
```diff
@@ -0,0 +1,14 @@
+part of 'sign_in_view_model.dart';
```
use base state

wasimGhoniem commented on lib/Feature/auth/presentation/viewModel/signin/sign_in_view_model.dart:
```diff
@@ -0,0 +1,28 @@
+        email: email,
+        password: password
+    );
+    result.when(
```
use api result class

YassenAhmed-gif committed:
Refactor: Enhance sign-in feature and remove unused screens

This commit refactors the sign-in functionality and removes the unused `ForgetPassword` and `SignUpScreen` files.

Key changes include:
- Updated `AuthRemoteDataSourceImpl` to use `SigninRequestEntity` and `SigninResponseEntity`.
- Added mock for `ApiServices` for testing purposes.
- Renamed `SigninResponseEntity` to `UserResponseEntity` and created a new `SigninResponseEntity`.
- Updated `SigninRequestDto` to use non-nullable types and added `toDto` method in `SigninRequestEntity`.
- Separated `UserDto` from `SigninResponseDto` into its own file.
- Updated `AuthRemoteDataSource` to use `SigninRequestEntity`.
- Updated `AuthRepoImpl` to use `SigninRequestEntity`.
- Created `UserDto` to represent user data transfer object.
- Updated `AuthRepo` to use `SigninRequestEntity`.
- Updated `SigninUseCase` to use `SigninRequestEntity`.
- Created `SigninRequestEntity` to represent sign-in request data.
- Updated `SigninViewModel` to manage form state and use `SigninRequestEntity`.
- Updated `SigninScreen` to use `SigninViewModel` for form handling and state management, and integrated validation logic.
- Removed `ForgetPassword` screen.
- Removed `SignUpScreen`.

YassenAhmed-gif committed:
Refactor: Remove unused mock API services

This commit removes the `api_services.dart` file from the `test/mocks` directory as it was no longer needed.

YassenAhmed-gif committed:
feat: Add tests for AuthRemoteDataSourceImpl

This commit introduces unit tests for the `AuthRemoteDataSourceImpl` class, covering the `signIn` method.

Key changes include:
- Added `auth_remote_data_source_impl_test.dart` with tests for successful API calls, DioException handling, generic exception handling, correct DTO mapping, and specific DioError response handling.
- Generated mock file `auth_remote_data_source_impl_test.mocks.dart` for `ApiServices`.
- Removed commented-out sections and simplified descriptions in `pubspec.yaml`.

ali-safwat commented on lib/Feature/auth/api/dataSources/auth_remote_data_source_impl.dart:
```diff
@@ -1,5 +1,33 @@
+  AuthRemoteDataSourceImpl(this._apiServices);
+
+  @override
+  Future<ApiResult<SigninResponseEntity>> signin({
```
for enhancment try use safeApi call to reduce try/catch

ali-safwat commented on lib/Feature/auth/data/repositories/auth_repo_impl.dart:
```diff
@@ -1,5 +1,44 @@
+    bool? rememberMeChecked = false
+  }) async {
+
+    // final ApiSuccessResult result = ApiSuccessResult(data: data);
```
remove this comment

ali-safwat commented on lib/Feature/auth/domain/use_cases/sign_in_use_case.dart:
```diff
@@ -0,0 +1,26 @@
+
+@injectable
+class SigninUseCase{
+  AuthRepo authRepo;
```
make this private more better to avoid any one access this repo throw use case

ali-safwat reviewed (approved):
done
